### ISSUE TITLE
Prover sometimes gives incorrect results for Connect Four

### ISSUE THREAD

**Issue by YenTheFirst** [gsingh93/ggp-rs#3]
using https://github.com/ggp-org/ggp-base/blob/master/games/games/connectFour/connectFour.kif,
The prover will sometimes give a nonsensical game state as the result of a sequence of moves.

demonstrating code

```
extern crate gdl_parser;
use gdl_parser::parse;
use gdl_parser::{Constant, Function};
extern crate ggp_rs;
use ggp_rs::{Role, Move, State};
use ggp_rs::prover::{Prover, query_builder};
use ggp_rs::util::create_does;

use std::fs::File;
use std::io::Read;

fn main() {
    let mut gdl = String::new();
    let mut gdl_file = File::open("connectFour.kif").ok().expect("unable to open file");
    gdl_file.read_to_string(&mut gdl).ok().expect("Unable to read file");

    let rules = parse(&gdl);
    let prover = Prover::new(rules);
    let init_state = prover.ask(query_builder::init_query(), &State::new()).into_state();
    println!("{}", init_state);

    let red = Role::new("red");
    let black = Role::new("black");
    let drop5 = Move::new(Function::new("drop", vec![Constant::new("5").into()]).into());
    let noop = Move::new(Constant::new("noop").into());
    let mut next_state_build = init_state.clone();
    next_state_build.props.insert(create_does(&red, &drop5));
    next_state_build.props.insert(create_does(&black, &noop));
    let after_move_1 = prover.ask(query_builder::next_query(), &next_state_build).into_state();
    println!("{}", after_move_1);
    let mut next_state_build = after_move_1.clone();
    next_state_build.props.insert(create_does(&red, &noop));
    next_state_build.props.insert(create_does(&black, &drop5));
    let after_move_2 = prover.ask(query_builder::next_query(), &next_state_build).into_state();
    println!("{}", after_move_2);
}
```

(this is using the most recent version of ggp-rs, though I locally forked in order to add a display function for states, and to mark certain functions public in order to write this test. Tested on rustc 1.1.0-nightly (7bd71637c 2015-05-06))

When run, this produces either the correct output 

```
(true (control red))
(true (cell 5 1 red)) (true (control black))
(true (cell 5 1 red)) (true (cell 5 2 black)) (true (control red))
```

or incorrect output

```
(true (control red))
(true (cell 5 1 red)) (true (control black))
(true (cell 5 1 black)) (true (cell 5 1 red)) (true (cell 5 2 black)) (true (control red))
```

Either output occurs with approximately 50% probability. I'm guessing the nondeterminism is related to HashMap randomization, but I haven't nailed it down yet.

**Comment by gsingh93** [comment]
Thanks for the report! I found a similar bug yesterday that I fixed last night. I figured it out after a few hours of debugging, and when I finished it was late so I decided to hold off on pushing the fix until today.

I'll clean up my patch, push it, and let you know here. I won't bump the version and release it just yet, as I found one more bug I want to look at before releasing, so once I push the fix you'll have to use `git = "..."` to use the project with the fix.
